File: src/game/components/WorldGenerator.ts
import { createNoise2D } from "simplex-noise";
import Component from "../engine/Component";
import TileMap from "../engine/components/TileMap";
import type { Vector2 } from "../engine/types";

type Address = `${number},${number}`;

class WorldGenerator extends Component {
	private seed: number = 0.5;
	private noise2D = createNoise2D(() => this.seed);
	private tileMapRef: TileMap | null = null;

	private readonly chunkSize: Vector2 = { x: 32, y: 32 };

	public setup(): void {
		this.tileMapRef = this.entity.getComponent(TileMap);
	}

	public update(_deltaTime: number): void {
		const camera = this.entity.game.getCamera();
		if (!camera) return;
		const bounds = camera.getBounds();
		if (!bounds) return;

		const [min, max] = bounds;

		const minChunkX = Math.floor(min.x / this.chunkSize.x) * this.chunkSize.x;
		const maxChunkX = Math.floor(max.x / this.chunkSize.x) * this.chunkSize.x;
		const minChunkY = Math.floor(min.y / this.chunkSize.y) * this.chunkSize.y;
		const maxChunkY = Math.floor(max.y / this.chunkSize.y) * this.chunkSize.y;

		for (let cx = minChunkX; cx <= maxChunkX; cx += this.chunkSize.x) {
			for (let cy = minChunkY; cy <= maxChunkY; cy += this.chunkSize.y) {
				this.generateChunk(cx, cy);
			}
		}
	}

	private generatedChunks: Set<Address> = new Set();
	private caveMap: Map<Address, boolean> = new Map();

	private generateChunk(cx: number, cy: number): void {
		if (!this.tileMapRef) return;
		const addr = this.encodeAddress(cx, cy);
		if (this.generatedChunks.has(addr)) return;

		// First pass: generate terrain and initial cave noise
		for (let dx = 0; dx < this.chunkSize.x; dx++) {
			const x = cx + dx;
			for (let dy = 0; dy < this.chunkSize.y; dy++) {
				const y = cy + dy;

				let noise = this.normalizedNoise(x / 50, this.seed);
				noise += this.normalizedNoise((x / 50) ** (1 / 2), this.seed) / 2;
				noise += this.normalizedNoise((x / 50) ** (1 / 4), this.seed) / 4;

				const landstripHeight = Math.floor(noise * 10);

				const stoneHeight =
					landstripHeight + this.normalizedNoise(x / 2, this.seed) * 4 + 5;

				if (y > stoneHeight) {
					this.tileMapRef.setTile(x, y, 3);
				} else if (y > landstripHeight) {
					this.tileMapRef.setTile(x, y, 1);
				} else if (y === landstripHeight) {
					this.tileMapRef.setTile(x, landstripHeight, 2);
				}

				// Enhanced cave generation with multiple frequencies
				const caveValue = this.getCaveValue(x, y);
				this.caveMap.set(this.encodeAddress(x, y), caveValue);
			}
		}

		// Second pass: apply cellular automata smoothing
		const smoothedCaves = this.smoothCaves(cx, cy);

		// Third pass: carve out caves
		for (let dx = 0; dx < this.chunkSize.x; dx++) {
			const x = cx + dx;
			for (let dy = 0; dy < this.chunkSize.y; dy++) {
				const y = cy + dy;
				const addr = this.encodeAddress(x, y);

				if (smoothedCaves.get(addr)) {
					this.tileMapRef.setTile(x, y, 0);
				}
			}
		}

		this.generatedChunks.add(addr);
	}

	private getCaveValue(x: number, y: number): boolean {
		// Layer 1: Large cave systems (scale 40)
		const largeCaves = this.normalizedNoise(x / 40, y / 40);

		// Layer 2: Medium tunnels (scale 20)
		const mediumCaves = this.normalizedNoise(x / 20, y / 20);

		// Layer 3: Small pockets (scale 10)
		const smallCaves = this.normalizedNoise(x / 10, y / 10);

		// Layer 4: Vertical variation using different seed offset
		const verticalNoise = this.normalizedNoise(x / 30, y / 30 + 1000);

		// Combine layers with weights
		const combined =
			largeCaves * 0.5 +
			mediumCaves * 0.3 +
			smallCaves * 0.15 +
			verticalNoise * 0.05;

		// Create caves where the combined value is below threshold
		// Lower threshold = more caves
		return combined < 0.42;
	}

	private smoothCaves(cx: number, cy: number): Map<Address, boolean> {
		const smoothed = new Map<Address, boolean>();

		// Apply cellular automata rules for more organic caves
		for (let dx = 0; dx < this.chunkSize.x; dx++) {
			const x = cx + dx;
			for (let dy = 0; dy < this.chunkSize.y; dy++) {
				const y = cy + dy;

				// Count cave neighbors in 3x3 grid
				let caveNeighbors = 0;
				for (let ox = -1; ox <= 1; ox++) {
					for (let oy = -1; oy <= 1; oy++) {
						if (ox === 0 && oy === 0) continue;
						const neighborAddr = this.encodeAddress(x + ox, y + oy);
						if (this.caveMap.get(neighborAddr)) {
							caveNeighbors++;
						}
					}
				}

				// Smoothing rules:
				// - If 5+ neighbors are caves, become a cave
				// - If 4- neighbors are caves, become solid
				const addr = this.encodeAddress(x, y);
				const isCave = caveNeighbors >= 5;
				smoothed.set(addr, isCave);
			}
		}

		return smoothed;
	}

	private normalizedNoise(x: number, y: number): number {
		const noiseValue = this.noise2D(x, y);
		return (noiseValue + 1) / 2; // Normalize to [0, 1]
	}

	private encodeAddress(x: number, y: number): Address {
		return `${Math.floor(x)},${Math.floor(y)}`;
	}
}

export default WorldGenerator;

File: rust-port/engine/src/components/world_generator.rs
use crate::component::{Component, RenderContext, SetupContext, UpdateContext};
use crate::impl_component_any;
use crate::types::{encode_address, Address, Vec2};
use noise::{NoiseFn, Simplex};
use std::collections::HashSet;

const CHUNK_SIZE: i32 = 32;

/// World generator component using Simplex noise
pub struct WorldGenerator {
    simplex: Simplex,
    seed: u32,
    generated_chunks: HashSet<Address>,
    /// Pending tiles to set: (x, y, tile_id)
    pub pending_tiles: Vec<(i32, i32, u8)>,
    /// Visible bounds for generation (set externally)
    pub visible_bounds: Option<(Vec2, Vec2)>,
}

impl Default for WorldGenerator {
    fn default() -> Self {
        Self::new(42)
    }
}

impl WorldGenerator {
    pub fn new(seed: u32) -> Self {
        Self {
            simplex: Simplex::new(seed),
            seed,
            generated_chunks: HashSet::new(),
            pending_tiles: Vec::new(),
            visible_bounds: None,
        }
    }

    pub fn set_visible_bounds(&mut self, min: Vec2, max: Vec2) {
        self.visible_bounds = Some((min, max));
    }

    /// Take pending tiles
    pub fn take_pending_tiles(&mut self) -> Vec<(i32, i32, u8)> {
        std::mem::take(&mut self.pending_tiles)
    }

    fn normalized_noise(&self, x: f64, seed_offset: u32) -> f64 {
        let value = self.simplex.get([x, seed_offset as f64]);
        (value + 1.0) / 2.0 // Normalize from [-1, 1] to [0, 1]
    }

    fn generate_chunk(&mut self, chunk_x: i32, chunk_y: i32) {
        let chunk_addr = encode_address(chunk_x, chunk_y);
        if self.generated_chunks.contains(&chunk_addr) {
            return;
        }
        self.generated_chunks.insert(chunk_addr);

        let world_x = chunk_x * CHUNK_SIZE;
        let world_y = chunk_y * CHUNK_SIZE;

        // First pass: terrain generation
        let mut tiles: Vec<(i32, i32, u8)> = Vec::new();

        for ty in 0..CHUNK_SIZE {
            for tx in 0..CHUNK_SIZE {
                let x = world_x + tx;
                let y = world_y + ty;

                // Calculate terrain height using noise
                let x_norm = x as f64 / 50.0;
                let mut noise = self.normalized_noise(x_norm, self.seed);
                noise += self.normalized_noise(x_norm.abs().sqrt() * x_norm.signum(), self.seed) / 2.0;
                noise += self.normalized_noise(x_norm.abs().powf(0.25) * x_norm.signum(), self.seed) / 4.0;

                let landstrip_height = (noise * 10.0).floor() as i32;
                let stone_height = landstrip_height + 5;

                // Determine tile type based on depth
                let tile_id = if y > stone_height {
                    // Below stone level - check for caves
                    let cave_noise = self.cave_noise(x, y);
                    if cave_noise < 0.42 {
                        0 // Cave (empty)
                    } else {
                        3 // Stone
                    }
                } else if y > landstrip_height {
                    // Dirt layer
                    1 // Dirt
                } else if y == landstrip_height {
                    // Surface
                    2 // Grass
                } else {
                    // Above ground
                    0 // Air
                };

                if tile_id > 0 {
                    tiles.push((x, y, tile_id));
                }
            }
        }

        // Second pass: cellular automata smoothing for caves
        // (simplified - full implementation would iterate multiple times)
        let smoothed = self.smooth_caves(&tiles, world_x, world_y);

        self.pending_tiles.extend(smoothed);
    }

    fn cave_noise(&self, x: i32, y: i32) -> f64 {
        let scale1 = 40.0;
        let scale2 = 20.0;
        let scale3 = 10.0;

        let n1 = self.simplex.get([x as f64 / scale1, y as f64 / scale1]);
        let n2 = self.simplex.get([x as f64 / scale2, y as f64 / scale2]);
        let n3 = self.simplex.get([x as f64 / scale3, y as f64 / scale3]);

        // Combine noise at different frequencies
        let combined = (n1 + 1.0) / 2.0 * 0.5 + (n2 + 1.0) / 2.0 * 0.3 + (n3 + 1.0) / 2.0 * 0.2;

        combined
    }

    fn smooth_caves(
        &self,
        tiles: &[(i32, i32, u8)],
        _chunk_x: i32,
        _chunk_y: i32,
    ) -> Vec<(i32, i32, u8)> {
        // Create a set for quick lookup
        let tile_set: HashSet<(i32, i32)> = tiles.iter().map(|(x, y, _)| (*x, *y)).collect();

        let mut result = Vec::new();

        for (x, y, tile_id) in tiles {
            // Count solid neighbors
            let mut solid_count = 0;
            for dy in -1..=1 {
                for dx in -1..=1 {
                    if dx == 0 && dy == 0 {
                        continue;
                    }
                    if tile_set.contains(&(x + dx, y + dy)) {
                        solid_count += 1;
                    }
                }
            }

            // Cellular automata rule: 5+ neighbors = solid
            if solid_count >= 5 || *tile_id == 2 {
                // Keep grass tiles always
                result.push((*x, *y, *tile_id));
            }
        }

        result
    }

    fn generate_visible_chunks(&mut self) {
        if let Some((min, max)) = self.visible_bounds {
            let chunk_min_x = (min.x as i32).div_euclid(CHUNK_SIZE) - 1;
            let chunk_max_x = (max.x as i32).div_euclid(CHUNK_SIZE) + 1;
            let chunk_min_y = (min.y as i32).div_euclid(CHUNK_SIZE) - 1;
            let chunk_max_y = (max.y as i32).div_euclid(CHUNK_SIZE) + 1;

            for chunk_y in chunk_min_y..=chunk_max_y {
                for chunk_x in chunk_min_x..=chunk_max_x {
                    self.generate_chunk(chunk_x, chunk_y);
                }
            }
        }
    }
}

impl Component for WorldGenerator {
    impl_component_any!(WorldGenerator);

    fn name(&self) -> &'static str {
        "WorldGenerator"
    }

    fn setup(&mut self, _ctx: &SetupContext) {}

    fn update(&mut self, _ctx: &UpdateContext) {
        // Generate chunks that are now visible
        self.generate_visible_chunks();
    }

    fn render(&self, _ctx: &mut RenderContext) {
        // WorldGenerator doesn't render
    }
}
